util/tracecpu/collector.go

```diff
@@ -0,0 +1,333 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracecpu
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tipb/go-tipb"
+	"github.com/wangjohn/quickselect"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+)
+
+// TopSQLRecord represents a single record of how much cpu time a sql plan consumes in one second.
+//
+// PlanDigest can be empty, because:
+// 1. some sql statements has no plan, like `COMMIT`
+// 2. when a sql statement is being compiled, there's no plan yet
+type TopSQLRecord struct {
+	SQLDigest  string
+	PlanDigest string
+	CPUTimeMs  uint32
+}
+
+type planBinaryDecodeFunc func(string) (string, error)
+
+// TopSQLDataPoints represents the cumulative SQL plan CPU time in current minute window
+type TopSQLDataPoints struct {
+	CPUTimeMsList  []uint32
+	TimestampList  []uint64
+	CPUTimeMsTotal uint64
+}
+
+type digestAndCPUTime struct {
+	Key            string
+	CPUTimeMsTotal uint64
+}
+type digestAndCPUTimeSlice []digestAndCPUTime
+
+func (t digestAndCPUTimeSlice) Len() int {
+	return len(t)
+}
+
+// We need find the kth largest value, so here should use >
+func (t digestAndCPUTimeSlice) Less(i, j int) bool {
+	return t[i].CPUTimeMsTotal > t[j].CPUTimeMsTotal
+}
+func (t digestAndCPUTimeSlice) Swap(i, j int) {
+	t[i], t[j] = t[j], t[i]
+}
+
+type planRegisterJob struct {
+	planDigest     string
+	normalizedPlan string
+}
+
+// TopSQLCollector is called periodically to collect TopSQL resource usage metrics
+type TopSQLCollector struct {
+	// // calling this can take a while, so should not block critical paths
+	// planBinaryDecoder planBinaryDecodeFunc
+	mu sync.RWMutex
+
+	// topSQLMap maps `sqlDigest-planDigest` to TopSQLDataPoints
+	topSQLMap map[string]*TopSQLDataPoints
+	maxSQLNum int
+
+	// normalizedSQLMap is an map, whose keys are SQL digest strings and values are normalized SQL strings
+	normalizedSQLMap map[string]string
+
+	// normalizedPlanMap is a plan version of normalizedSQLMap
+	// this should only be set from the dedicated worker
+	normalizedPlanMap map[string]string
+
+	planRegisterChan chan *planRegisterJob
+
+	// current tidb-server instance ID
+	instanceID string
+
+	agentGRPCAddress string
```

Comment:
The agentGRPCAddress (and other configurations) are dynamic, and may be changed at any time. @crazycs520 PTAL

Reply:
If we need to dynamically changing `agentGRPCAddress`, I should add an Public API to `TopSQLCollector`, e.g. `SetAgentGRPCAddress()`, and will take effect at the next send time.

Reply:
Added `SetAgentGRPCAddress()` API